### PR TITLE
PB-351 Store participant metrics

### PR DESCRIPTION
**Pull request by Robert-Steiner**
### References

[PB-351 Implement metric proxying/re-routing on coordinator for metrics shipped by participant](https://xainag.atlassian.net/browse/PB-351)

### Summary

* store participant metrics on InfluxDB
* InfluxDBClient is configurable via the config.toml
* feature is enabled in `docker-compose-dev` only

### Are there any open tasks/blockers for the ticket?

* Adjust the metrics messages to meet the needs of InfluxDB.(PB-390)

---

### Reviewer checklist

*Reviewer agreement:*

* Reviewers assign themselves at the start of the review.
* Reviewers do **not** commit or merge the merge request.
* Reviewers have to check and mark items in the checklist.

**Merge request checklist**

- [ ] Conforms to the merge request title naming `XP-XXX <a description in imperative form>`.
- [ ] Each commit conforms to the naming convention `XP-XXX <a description in imperative form>`.
- [ ] Linked the ticket in the merge request title or the references section.
- [ ] Added an informative merge request summary.

**Code checklist**

- [ ] Conforms to the branch naming `XP-XXX-<a_small_stub>`.
- [ ] Passed scope checks.
- [ ] Added or updated tests if needed.
- [ ] Added or updated code documentation if needed.
- [ ] Conforms to Google docstring style.
- [ ] Conforms to XAIN structlog style.

### CODE REVIEW

File: xain_fl/config/schema.py [hunk 1]
```diff
@@ -229,11 +229,27 @@ def validate_path(path):
...
 )
 
+METRICS_SCHEMA = Schema(
+    {
+        "host": And(
+            str,
+            hostname_or_ip_address,
+            error=error("metrics.host", "a valid hostname or ip address"),
+        ),
+        "port": Use(int, error=error("metrics.port", "a valid port number")),
```
Comment (little-dude) [inline xain_fl/config/schema.py]: I think there's a "positive integer" validator in that module already

File: xain_fl/config/schema.py [hunk 2]
```diff
@@ -268,6 +284,9 @@ def create_class_from_schema(class_name: str, schema: Schema) -> Any:
 StorageConfig = create_class_from_schema("StorageConfig", STORAGE_SCHEMA)
 StorageConfig.__doc__ = "Storage related configuration: storage endpoints and credentials, etc."
 
+MetricsConfig = create_class_from_schema("MetricsConfig", METRICS_SCHEMA)
+MetricsConfig.__doc__ = "Storage related configuration: storage endpoints and credentials, etc."
```
Comment (little-dude) [inline xain_fl/config/schema.py]: This docstring needs to be updated

File: xain_fl/coordinator/metrics_store.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        Args:
+
+        participant_ids: The list of IDs of the participants selected
+            to participate in this round.
+        metrics :
+        """
+
+
+class NoMetricsStore(ABCMetricsStore): # pylint: disable=too-few-public-methods
```
Comment (little-dude) [inline xain_fl/coordinator/metrics_store.py]: In `xain-sdk` I called the storage counterpart of this class `DummyStore`. Maybe we should agree on a naming convention for these classes that do nothing.

File: xain_fl/config/schema.py [hunk 4]
```diff
@@ -154,51 +153,23 @@ def hostname_or_ip_address(
     return And(str, Or(is_valid_hostname, is_valid_ip_address), error=error(key, expected_value),)
 
 
-def abs_path(value: str) -> str:
```
Comment (little-dude) [inline xain_fl/config/schema.py]: Why is this removed? Isn't it used?

File: xain_fl/config/schema.py [hunk 5]
```diff
@@ -229,28 +200,36 @@ def validate_path(path):
...
+        Optional("password", default="root"): Use(
+            str, error=error("metrics.password", "a valid password")
+        ),
+        Optional("db_name", default="metrics"): Use(
+            str, error=error("metrics.db_name", "a database name")
+        ),
+    }
+)
 
-LOGGING_SCHEMA = Schema({Optional("level", default="info"): log_level("logging.level"),})
```
Comment (little-dude) [inline xain_fl/config/schema.py]: I think this was removed by mistake

File: xain_fl/coordinator/metrics_store.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        into a metric store.
+
+        Args:
+
+            participant_id: The ID of the participant.
+            metrics: The metrics of the participant with the given participant_id.
+
+        Returns:
+
+            True, on success, otherwise False.
```
Comment (little-dude) [inline xain_fl/coordinator/metrics_store.py]: I feel like it's a bit weird in python to use a boolean for reporting a result. Do you think we could use a custom `MetricsStoreError` exception instead?

File: xain_fl/coordinator/metrics_store.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        """
+
+        # FIXME: We will change the data format of the metrics message in a separate ticket.
+        # The goal is, that coordinator doesn't need to transform the metrics anymore.
```
Comment (little-dude) [inline xain_fl/coordinator/metrics_store.py]: I'm not sure to understand this comment: does that means that the `metrics` argument will be a dictionary with the right structure and that the transformation will be done by the caller?

File: xain_fl/coordinator/metrics_store.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            logger.warn("Can not write metrics", error=str(err))
+            return False
+
+
+def transform_metrics_to_influx_data_points(
+    participant_id: str, metrics: Dict[str, ndarray]
+) -> List[dict]:
+    """Transform the metrics of a participant into InfluxDB data points.
+
+    Arguments:
```
Comment (little-dude) [inline xain_fl/coordinator/metrics_store.py]: `Arguments` should be `Args`

File: xain_fl/coordinator/metrics_store.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        participant_id: The ID of the participant.
+        metrics: The metrics of the participant with the given participant_id.
+
+    Returns:
+        The metrics of the participant as InfluxDB data points.
+    """
+
+    # Currently the metrics message does not contain any timestamps.
+    # We set a timestamp for each epoch data point with an interval of 1 sec.
```
Comment (little-dude) [inline xain_fl/coordinator/metrics_store.py]: What is the reason for this?

File: xain_fl/coordinator/metrics_store.py (hunk 7 above)
Comment (Robert-Steiner) [reply xain_fl/coordinator/metrics_store.py]: Right. The idea is, that a participant sends the metrics already in s influx compatible data schema so that the coordinator just has to pass the data to InfluxDB without any transformation of the data.

File: xain_fl/coordinator/metrics_store.py (hunk 9 above)
Comment (Robert-Steiner) [reply xain_fl/coordinator/metrics_store.py]: Since we are using a time series database, we need to include a timestamp for each data point. The interval thing is just for the case that the participant send metrics for more than 1 epoch per round. If we just use the same timestamp, all data points will group on a horizontal line.

File: xain_fl/coordinator/metrics_store.py (hunk 6 above)
Comment (Robert-Steiner) [reply xain_fl/coordinator/metrics_store.py]: Good point, I will adjust it tomorrow.

File: xain_fl/coordinator/metrics_store.py (hunk 9 above)
Comment (little-dude) [reply xain_fl/coordinator/metrics_store.py]: That makes sense, I think it is worth adding an explanation as a comment.

Comment (Robert-Steiner) [reply xain_fl/coordinator/metrics_store.py]: I added it to the comment

File: xain_fl/coordinator/metrics_store.py (hunk 6 above)
Comment (Robert-Steiner) [reply xain_fl/coordinator/metrics_store.py]: I have adjusted the code
